### ISSUE TITLE
Association quarantined after new incarnation

### ISSUE THREAD

**Issue by patriknw** [akka/akka#30054]
Hi @patriknw, We are facing exactly similar issue. Only difference in our case is that, we  are using remote actorRef with `use-unsafe-remote-features-outside-cluster = on` and not using `actorSelection`.

Tried versions:
Akka typed - `2.6.11`, `2.6.12`

We have following setup:

1. `Actor1` => Running in one JVM and it has registered its remote actorRef in our custom `LocationService`, consider this similar to `Receptionist`

2. `Actor2` => Running in second JVM and registered its address with `LocationService`

Scenario:
1. `Actor1` sends `Restart` message to `Actor2` in loop `2000` times.
2. On `Restart` message `Actor2` terminates its `ActorSystem` and starts new on random port (`port=0`)
3. This works fine for 1600-1700 iterations
4. But eventually fails with following message:
```
Dropping message [esw.ocs.api.actor.messages.SequencerMessages$GetSequenceComponent] from
[Actor[akka://sequence-manager/deadLetters]] to 
[Actor[akka://sequencer-system@192.168.1.3:50944/user/ESW.Perf#1625616268]] 
due to quarantined system [akka://sequencer-system@192.168.1.3:50944]
```

Here,
sequence-manager = Actor1
sequencer-system = Actor2

Let me know if you would like to take a look at detailed log file, its 150MB in size hence not attaching here.

_Originally posted by @kpritam in https://github.com/akka/akka/issues/29828#issuecomment-782572479_

**Comment by patriknw** [comment]
See logs in https://github.com/akka/akka/issues/29828#issuecomment-783160509

**Comment by patriknw** [comment]
Extract of the logs:
```
2021-02-19T11:44:58.343Z INFO :slf4j  (Slf4jLogger.scala 96) - Remoting started with transport [Artery tcp]; listening on address [akka://sequencer-system@192.168.1.3:50944] with UID [2832747598290303135]
2021-02-19T11:44:58.380Z INFO  CSW.location_service_lib (LocationServiceImpl.scala 49) - Registering connection: [ESW.Perf-Sequencer-akka] with location: [akka://sequencer-system@192.168.1.3:50944/user/ESW.Perf#1844042501]
2021-02-19T11:44:58.414Z INFO :akka  - Cluster Node [akka://sequencer-system@192.168.1.3:50944] - Successfully shut down
2021-02-19T11:44:59.417Z INFO :akka  - Association to [akka://sequencer-system@192.168.1.3:50944] having UID [2832747598290303135] has been stopped. All messages to this UID will be delivered to dead letters. Reason: ActorSystem terminated


2021-02-19T11:46:22.012Z INFO :akka  - Remoting started with transport [Artery tcp]; listening on address [akka://sequencer-system@192.168.1.3:50944] with UID [-6403714036655331529]
2021-02-19T11:46:22.057Z INFO  CSW.location_service_lib (LocationServiceImpl.scala 82) - Successfully registered connection: [ESW.Perf-Sequencer-akka] with location [akka://sequencer-system@192.168.1.3:50944/user/ESW.Perf#1625616268]
2021-02-19T11:46:22.060Z DEBUG:akka  - Dropping message [esw.ocs.api.actor.messages.SequencerMessages$GetSequenceComponent] from [Actor[akka://sequence-manager/deadLetters]] to [Actor[akka://sequencer-system@192.168.1.3:50944/user/ESW.Perf#1625616268]] due to quarantined system [akka://sequencer-system@192.168.1.3:50944]
```

It looks like the association is still considered as quarantined even though a new incarnation has connected. I find that strange since there should have been a handshake for the register message of akka://sequencer-system@192.168.1.3:50944/user/ESW.Perf#1625616268

Maybe some race condition. I guess GetSequenceComponent is a reply to the register message? The ActorRef ESW.Perf#1625616268 is the new one, that was registered.

**Comment by kpritam** [comment]
> Maybe some race condition. I guess GetSequenceComponent is a reply to the register message? The ActorRef ESW.Perf#1625616268 is the new one, that was registered.

Complete flow looks like below (code that is running in loop)
```scala
sequenceManagerRef ! Restart
  - sequencerRef = lookup(CRDT's)
  - sequenceComponentRef = sequencerRef ? GetSequenceComponent
  - sequenceComponentRef ! Restart
    - Stop sequencerRef and shutdown sequencer ActorSystem
    - Start new sequencer ActorSystem
    - Start sequencerRef -> register (add to CRDT's)
```    

Here the last step is to register `sequencerRef` with CRDT's. And when I look at iteration wise logs,
this call `lookup(CRDT's)`, always gives me ref registered in previous iteration's last step. Hence does not look like I am getting stale reference. I also printed out state of CRDT when test fails, it does print same `actorRef` for which we are getting `Dropping message ...` error log

**Comment by patriknw** [comment]
Is the register message sent with actorSelection from the new sequencer-system to the LocationServiceImpl in sequence-manager?

**Comment by kpritam** [comment]
> Is the register message sent with actorSelection from the new sequencer-system to the LocationServiceImpl in sequence-manager?

No, We have our own implementation of `LocationService` which stores all the serialized actor ref locations in the CRDT using something similar to below:
```scala
implicit class RichActor(actorRef: ActorRef[_]) {
    def toURI: URI = new URI(Serialization.serializedActorPath(actorRef.toClassic))
  }

implicit class RichURI(val uri: URI) {
    def toActorRef(implicit system: ActorSystem[_]): ActorRef[Nothing] = ActorRefResolver(system).resolveActorRef(uri.toString)
  }
```

**Note:** `LocationService` provides HTTP interface to lookup registered actorRef's locations. 
Every machine starts one instance of `LocationServer` which takes part in forming Akka Cluster.
Other JVM's on that machine requests for location of named refs or registers new location via localhost location http sever.

**Comment by patriknw** [comment]
Ah, so the registration is not via an actor message from sequencer-system.
```
2021-02-19T11:44:58.380Z INFO  CSW.location_service_lib (LocationServiceImpl.scala 49) - Registering connection: [ESW.Perf-Sequencer-akka] with location: [akka://sequencer-system@192.168.1.3:50944/user/ESW.Perf#1844042501]
```

So the very first remote message is the GetSequenceComponent from sequence-manager to sequencer-system

**Comment by kpritam** [comment]
Yes, that is correct. Register is HTTP call. And `GetSequenceComponent` is first remote actor msg.

**Comment by patriknw** [comment]
I can reproduce this problem in a test. It can't send the message to the quarantined system when there is nothing else that is breaking the quarantine. No incoming message, and no actorSelection.

What you can do is to use actorSelection before the first message to a new incarnation. If that is successful it will break the quarantine and the ActorRef can be used.

You can send `ActorIdentity` with actorSelection to any path of the sequencer-system before sending the GetSequenceComponent. Best to wait for the ActorIdentity reply before sending the GetSequenceComponent message.
In logs you will then see how the incarnation is increased:
```
Incarnation 2 of association to [akka://systemB@localhost:62163] with new UID [2299450648948447906] (old UID [2006392835401071027])
```

This is what I used in my test:
```
system.actorSelection(RootActorPath(address(systemC)) / "user").tell(Identify("piercing"), piercingProbe.ref)
piercingProbe.expectMsgType[ActorIdentity](1.seconds).ref.get
          
val echo2Ref = system.asInstanceOf[ExtendedActorSystem].provider.resolveActorRef(serializedEcho2)
echo2Ref ! "ping2"
expectMsg(5.seconds, "pong2")
```

**Comment by shivanithorat** [comment]
In the above application mentioned by @kpritam where sequencer actor-system(actor-system 1) getting quarantined, we are observing that memory heap of sequencer-manager process(actor-system 2) is increasing. Is this increase due to all the quarantined associations being stored in the sequence manager process.  We have also used `akka.remote.artery.advanced.remove-quarantined-association-after = 10s` flag in the application.conf of sequence manager process.

**Comment by shivanithorat** [comment]
I am attaching a few [logs](https://drive.google.com/file/d/1niDiLhRJOwObCSX-PRAVv63aHRWhwJAt/view?usp=sharing) of a short experiment that will depict the situation mentioned above. The logs are taken at a time interval of 2.5 hours and we can see an increase in the heap size of about 7.9 MB. We have also used `akka.remote.artery.advanced.remove-quarantined-association-after = 10s` flag in the application.conf of sequence manager process. When we perform load testing for 10 hours with similar setup we see an increase in the heap size of 30 MB.

**Comment by patriknw** [comment]
@shivanithorat Thanks for those heap dumps. That's very useful. I have looked at them and what is clear is that about 1660 instances of `Association` are strongly reachable. Those are rather lightweight (since connections and streams are stopped) but still remains in the heap. Looking at incoming references it seems like many are from the ActorRefResolveCache which is used in serialization via a thread local or from Decoder in Artery.

Those caches are bounded, by hardcoded capacity of 1024, but in total there could be many references if the serialization caches are accessed from many threads.

Would be nice if we could evict those caches and remove entries of RemoteActorRef that have cachedAssociation to Association that has been removed after quarantined.
